--- gcc/read-rtl.c.orig	2018-04-22 22:37:39.586820722 -0400
+++ gcc/read-rtl.c	2018-04-22 22:43:54.055417498 -0400
@@ -659,7 +659,7 @@ again:
 	    {
 	      ungetc (c, infile);
 	      list_counter++;
-	      obstack_ptr_grow (&vector_stack, (PTR) read_rtx (infile));
+              obstack_ptr_grow (&vector_stack, read_rtx (infile));
 	    }
 	  if (list_counter > 0)
 	    {
--- gcc/system.h.orig	2018-04-22 22:37:39.586820722 -0400
+++ gcc/system.h	2018-04-22 22:46:08.683621593 -0400
@@ -589,7 +589,9 @@ typedef char _Bool;
 #undef realloc
 #undef calloc
 #undef strdup
+#if ! (defined(FLEX_SCANNER) || defined(YYBISON) || defined(YYBYACC))
  #pragma GCC poison malloc realloc calloc strdup
+#endif

 /* Old target macros that have moved to the target hooks structure.  */
  #pragma GCC poison ASM_OPEN_PAREN ASM_CLOSE_PAREN			\
@@ -628,4 +630,11 @@ typedef char _Bool;

 #endif /* GCC >= 3.0 */

+#if defined(FLEX_SCANNER) || defined(YYBISON) || defined(YYBYACC)
+/* Flex and bison use malloc and realloc.  Yuk.  Note that this means
+   really_call_* cannot be used in a .l or .y file.  */
+#define malloc xmalloc
+#define realloc xrealloc
+#endif
+
 #endif /* ! GCC_SYSTEM_H */
